proposals/0000-fine-grained-undecidable.md

```diff
@@ -0,0 +1,185 @@
+
+```haskell
+data Pair a = Pair (a, a)
+instance Show (a, a) => Show (Pair a) where ...
+```
+
+As GHC's error message point out, one can escape those checks by using `UndecidableInstances`.
+
+```
+<interactive>: error:
+    • Variable ‘a’ occurs more often
+        in the constraint ‘Show (a, a)’
+        than in the instance head ‘Show (Pair a)’
+      (Use UndecidableInstances to permit this)
+```
+
+However, this is quite a drastic measure. In order to accept _one_ instance, we need to disable the checks for _every_ instance defined in the module. For example, now looping instances are allowed without warning.
+
+```haskell
+instance Show (Pair a) => Show (Pair a)
+```
+
+### The many faces of `Undecidable`
+
+As it turns out, GHC puts a lot of tweak under the `UndecidableSuperClasses` and `UndecidableInstances` extensions. A few of them are related to termination proper:
+
+- `UndecidableSuperClasses` omits the cycle check for all class declarations in a module. This allows for arbitrary predicates in a superclass constraint.
+- `UndecidableInstances` omits the termination check for all local quantified constraints in a module.
+- `UndecidableInstances` skips the termination check for class instances.
+- `UndecidableInstances` skips the termination check for type family equations (both in open type family instances and closed type family equations).
+
+There are also tweaks to the coverage conditions for coverage / injectivity in type families and functional dependencies.
+
+- `UndecidableInstances` allows a liberalization of the check for injective type family consistency.
+- `UndecidableInstances` allows an instance of a class with functional dependencies to use the liberal coverage condition instead of the coverage condition.
+
+## Proposed Change Specification
+
+We introduce new modifiers `%NoTerminationCheck`, `%LiberalCoverage`, `%LiberalInjectivity`, `%Overlapping`, `%Overlappable`, and `%Overlaps` which are used _before_ the keyword introducing the definition where we want to lift the corresponding restriction. The modifier syntax is taken from this [approved proposal](https://github.com/ghc-proposals/ghc-proposals/pull/370), with the additional [change](https://github.com/ghc-proposals/ghc-proposals/pull/392) of modifiers appearing _before_ the modified thing.
+
+1. Putting `%NoTerminationCheck` before a class declaration skips its superclass-cycle check.
+
+    ```haskell
+    %NoTerminationCheck ; class C (F a) => C a where ...
+    ```
+
+    ```diff
+    +cls_mod  : '%NoTerminationCheck'
+    +cls_mods : {- empty -} | cls_mods cls_mod
+
+    cl_decl :
+    - : 'class' tycl_hdr fds where_cls
+    + : cls_mods ';' 'class' tycl_hdr fds where_cls
```

Comment:
This can't be right.  It suggests that if there are no `cls_mods` then the class decl still needs to be preceded by a semi-colon.